akka-contrib/src/main/scala/akka/contrib/circuitbreaker/CircuitBreakerProxy.scala

```diff
@@ -0,0 +1,258 @@
+    circuitEventListener: Option[ActorRef] = None,
+    failureDetector: Any ⇒ Boolean = { _ ⇒ false },
+    openCircuitFailureConverter: CircuitOpenFailure ⇒ Any = identity) {
+
+    def withMaxFailures(value: Int) = copy(maxFailures = value)
+    def withCallTimeout(value: Timeout) = copy(callTimeout = value)
+    def withResetTimeout(value: Timeout) = copy(resetTimeout = value)
+    def withCircuitEventListener(value: Option[ActorRef]) = copy(circuitEventListener = value)
+    def withFailureDetector(value: Any ⇒ Boolean) = copy(failureDetector = value)
+    def withOpenCircuitFailureConverter(value: CircuitOpenFailure ⇒ Any) = copy(openCircuitFailureConverter = value)
+
+    /**
+     * Creates the props for a [[akka.contrib.circuitbreaker.CircuitBreakerProxy]] proxying the given target
+     *
+     * @param target the target actor ref
+     */
+    def props(target: ActorRef) = CircuitBreakerProxy.props(target, maxFailures, callTimeout, resetTimeout, circuitEventListener, failureDetector, openCircuitFailureConverter)
+
+  }
+}
+
+object CircuitBreakerInternalEvents {
+  sealed trait CircuitBreakerInternalEvent
+  case object CallFailed extends CircuitBreakerInternalEvent
+  case object CallSucceeded extends CircuitBreakerInternalEvent
+}
+
+import akka.contrib.circuitbreaker.CircuitBreakerInternalEvents._
+import akka.contrib.circuitbreaker.CircuitBreakerProxy._
+
+final class CircuitBreakerProxy(
+  target: ActorRef,
+  maxFailures: Int,
+  callTimeout: Timeout,
+  resetTimeout: Timeout,
+  circuitEventListener: Option[ActorRef],
+  failureDetector: Any ⇒ Boolean,
+  failureMap: CircuitOpenFailure ⇒ Any) extends Actor with ActorLogging with FSM[CircuitBreakerState, CircuitBreakerStateData] {
+
+  context watch target
+
+  startWith(Closed, CircuitBreakerStateData(failureCount = 0))
+
+  def callSucceededHandling: StateFunction = {
+    case Event(CallSucceeded, state) ⇒
+      log.debug("Received call succeeded notification in state {} resetting counter", state)
+      goto(Closed) using CircuitBreakerStateData(failureCount = 0, firstHalfOpenMessageSent = false)
+  }
+
+  def passthroughHandling: StateFunction = {
+    case Event(Passthrough(message), state) ⇒
+      log.debug("Received a passthrough message in state {}, forwarding the message to the target actor without altering current state", state)
+      target ! message
+      stay
+  }
+
+  def targetTerminationHandling: StateFunction = {
+    case Event(Terminated(`target`), state) ⇒
+      log.debug("Target actor {} terminated while in state {}, terminating this proxy too", target, state)
+      stop
+  }
+
+  def commonStateHandling: StateFunction = { callSucceededHandling orElse passthroughHandling orElse targetTerminationHandling }
+
+  when(Closed) {
+    commonStateHandling orElse {
+      case Event(TellOnly(message), _) ⇒
+        log.debug("CLOSED: Sending message {} without expecting any response", message)
```

Comment:
Inconsistency-nitpick: the state is called Closed but log says CLOSED, same for debug logs below.